In [2]:
import pandas as pd
import numpy as np
import math
np.random.seed(123)
# Create list of values for names column.

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# Randomly generate arrays of scores for each student for each subject.
# Note that all the values need to have the same length here.

math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))
# Construct the DataFrame using the above lists and arrays.

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades,
                   'classroom': np.random.choice(['A', 'B'], len(students))})

In [3]:
from pydataset import data

In [4]:
def get_db_url(db_name):
    from env import user, host, password
    return f'mysql+pymysql://{user}:{password}@{host}/{db_name}'

In [185]:
url = get_db_url("employees")
sql = '''
SELECT
    *
FROM employees
'''

employees = pd.read_sql(sql, url)
employees

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

### Once you have successfully run a query:

a. Intentionally make a typo in the database url. What kind of error message do you see?
    
    `NameError: name 'get_3_url' is not defined`

b. Intentionally make an error in your SQL query. What does the error message look like?

    `OperationalError: (pymysql.err.OperationalError) (1054, "Unknown column 'emp_number' in 'field list'")[SQL: SELECT emp_number FROM employees]`

In [186]:
url = get_db_url("employees")
sql = '''
SELECT
    *
FROM titles
'''

titles = pd.read_sql(sql, url)
titles

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

### 6-7. Read the employees and titles tables into two separate DataFrames. How many rows and columns do you have in each DataFrame? Is that what you expected?
* employees : 300024 x 6
* titles: 443308 x 4

### 8. Display the summary statistics for each DataFrame.

In [7]:
employees.info()
employees.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300024 entries, 0 to 300023
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   emp_no      300024 non-null  int64 
 1   birth_date  300024 non-null  object
 2   first_name  300024 non-null  object
 3   last_name   300024 non-null  object
 4   gender      300024 non-null  object
 5   hire_date   300024 non-null  object
dtypes: int64(1), object(5)
memory usage: 13.7+ MB


emp_no
count  300024.000000
mean   253321.763392
std    161828.235540
min     10001.000000
25%     85006.750000
50%    249987.500000
75%    424993.250000
max    499999.000000

In [8]:
titles.info()
titles.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443308 entries, 0 to 443307
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   emp_no     443308 non-null  int64 
 1   title      443308 non-null  object
 2   from_date  443308 non-null  object
 3   to_date    443308 non-null  object
dtypes: int64(1), object(3)
memory usage: 13.5+ MB


emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000

### 9. How many unique titles are in the titles DataFrame?



In [9]:
url = get_db_url("employees")
sql = '''
SELECT DISTINCT
    title
FROM titles
'''

titles = pd.read_sql(sql, url)
titles.shape

(7, 1)

### 10. What is the oldest date in the to_date column?



In [10]:
url = get_db_url("employees")
sql = '''
SELECT
    to_date
FROM titles
ORDER BY to_date ASC
'''

titles = pd.read_sql(sql, url)
titles.head(1)

to_date
0  1985-03-01

### 11. What is the most recent date in the to_date column?

In [11]:
url = get_db_url("employees")
sql = '''
SELECT
    to_date
FROM titles
ORDER BY to_date DESC
'''

titles = pd.read_sql(sql, url)
titles.head(1)

to_date
0  9999-01-01

## Exercises II
1. Copy the users and roles DataFrames from the examples above.

In [12]:
roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles

id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

In [13]:
users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users

id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

### 2. What is the result of using a right join on the DataFrames?


In [14]:
users.merge(roles, left_on='role_id', right_on='id', how='right', indicator=True)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0     1      admin        both
1   2.0    joe      2.0     2     author        both
2   3.0  sally      3.0     3   reviewer        both
3   4.0   adam      3.0     3   reviewer        both
4   NaN    NaN      NaN     4  commenter  right_only

### 3. What is the result of using an outer join on the DataFrames?

In [15]:
users.merge(roles, left_on='role_id', right_on='id', how='outer', indicator=True)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0   1.0      admin        both
1   2.0    joe      2.0   2.0     author        both
2   3.0  sally      3.0   3.0   reviewer        both
3   4.0   adam      3.0   3.0   reviewer        both
4   5.0   jane      NaN   NaN        NaN   left_only
5   6.0   mike      NaN   NaN        NaN   left_only
6   NaN    NaN      NaN   4.0  commenter  right_only

### 4. What happens if you drop the foreign keys from the DataFrames and try to merge them?


In [16]:
users.merge(roles, indicator=True)

Empty DataFrame
Columns: [id, name, role_id, _merge]
Index: []

### 5. Load the mpg dataset from PyDataset.


In [17]:
from pydataset import data
mpg = data('mpg')
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

### 6. Output and read the documentation for the mpg dataset.


In [18]:
mpg.info()
mpg.describe()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


displ         year         cyl         cty         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

### 7. How many rows and columns are in the dataset?


In [19]:
mpg.shape

(234, 11)

### 8. Check out your column names and perform any cleanup you may want on them.


In [20]:
mpg.rename(columns = {"cty": "city"}, inplace = True)
mpg.rename(columns = {"hwy": "highway"}, inplace = True)
mpg.loc[0:10,'city':'class']


city  highway fl    class
1     18       29  p  compact
2     21       29  p  compact
3     20       31  p  compact
4     21       30  p  compact
5     16       26  p  compact
6     18       26  p  compact
7     18       27  p  compact
8     18       26  p  compact
9     16       25  p  compact
10    20       28  p  compact

### 9. Display the summary statistics for the dataset.


In [21]:
mpg.describe()

displ         year         cyl        city     highway
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

### 10. How many different manufacturers are there?


In [98]:
how_many_manu = mpg.groupby('manufacturer').agg(['mean']).index
print(how_many_manu.shape)

(15,)


### 11. How many different models are there?


In [23]:
how_many_models = mpg.groupby('model').agg(['mean']).index
print(how_many_models.shape)

(38,)


### 12. Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.


In [24]:
mpg['mileage_difference'] = mpg['highway'] - mpg['city']
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  
1    compact                  11  
2    compact                   8  
3    compact                  11  
4    compact                   9  
5    compact                  10  
..       ...                 ...  
230  midsize                   9  
231  midsize                   8  
232  midsize                  10  
233  midsize                   8  
234  midsize                   9  

[234 rows x 12 columns]

### 13. Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.


In [99]:
mpg['average_mileage'] = (mpg['highway'] + mpg['city'])/2
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  average_mileage  is_automatic  
1    compact                  11             23.5          True  
2    compact                   8             25.0         False  
3    compact                  11             25.5         False  
4    compact                   9             25.5          True  
5    compact                  10             21.0          True  
..       ...                 ...              ...           ...  
230  midsize                   9             23.5          True  
231  midsize                   8             25.0         False  
232  midsize                  10             21.0          True  
233  midsize                   8             22.0         False  
234  midsize                   9             21.5          True  

[234 rows x 14 columns]

### 14. Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission.


In [80]:
mpg['is_automatic'] = mpg.trans.str.startswith('auto')
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  average_mileage  is_automatic  
1    compact                  11             23.5          True  
2    compact                   8             25.0         False  
3    compact                  11             25.5         False  
4    compact                   9             25.5          True  
5    compact                  10             21.0          True  
..       ...                 ...              ...           ...  
230  midsize                   9             23.5          True  
231  midsize                   8             25.0         False  
232  midsize                  10             21.0          True  
233  midsize                   8             22.0         False  
234  midsize                   9             21.5          True  

[234 rows x 14 columns]

### 15. Using the mpg dataset, find out which which manufacturer has the best miles per gallon on average?


In [192]:
best_miles_per_gallon = mpg.average_mileage.max()
best_miles_per_gallon

mpg.sort_values(by='average_mileage', ascending = False).head(1)

manufacturer       model  displ  year  cyl       trans drv  city  highway  \
222   volkswagen  new beetle    1.9  1999    4  manual(m5)   f    35       44   

    fl       class  mileage_difference  average_mileage  is_automatic  
222  d  subcompact                   9             39.5         False

### 16. Do automatic or manual cars have better miles per gallon?

In [82]:
mpg[mpg.trans.str.startswith('auto')].average_mileage.median()

19.0

In [83]:
mpg[mpg.trans.str.startswith('manual')].average_mileage.median()

22.0

# Exercises III

### 1. Use your get_db_url function to help you explore the data from the chipotle database.

In [132]:
orders = pd.read_sql('SELECT * FROM orders', get_db_url('chipotle'))


### 2. What is the total price for each order?


In [139]:
item_price = orders.item_price.str.replace("$", "").astype(float)
type(item_price)

<ipython-input-139-189dcf36c31f>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  item_price = orders.item_price.str.replace("$", "").astype(float)


pandas.core.series.Series

In [145]:
orders.groupby('order_id').agg('sum')


id  quantity
order_id                 
1            10         4
2             5         2
3            13         2
4            17         2
5            21         2
...         ...       ...
1830       9223         2
1831      13842         3
1832       9233         2
1833       9237         2
1834      13863         3

[1834 rows x 2 columns]

In [146]:
orders

id  order_id  quantity                              item_name  \
0        1         1         1           Chips and Fresh Tomato Salsa   
1        2         1         1                                   Izze   
2        3         1         1                       Nantucket Nectar   
3        4         1         1  Chips and Tomatillo-Green Chili Salsa   
4        5         2         2                           Chicken Bowl   
...    ...       ...       ...                                    ...   
4617  4618      1833         1                          Steak Burrito   
4618  4619      1833         1                          Steak Burrito   
4619  4620      1834         1                     Chicken Salad Bowl   
4620  4621      1834         1                     Chicken Salad Bowl   
4621  4622      1834         1                     Chicken Salad Bowl   

                                                                     choice_description  \
0                                                                                   nan   
1                                                                          [Clementine]   
2                                                                               [Apple]   
3                                                                                   nan   
4            [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   
...                                                                                 ...   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour Cream, Cheese, Lettuce, Guacamole]]   
4618               [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese, Lettuce, Guacamole]]   
4619         [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Guacamole, Lettuce]]   
4620                                 [Fresh Tomato Salsa, [Fajita Vegetables, Lettuce]]   
4621                    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Lettuce]]   

     item_price  
0        $2.39   
1        $3.39   
2        $3.39   
3        $2.39   
4       $16.98   
...         ...  
4617    $11.75   
4618    $11.75   
4619    $11.25   
4620     $8.75   
4621     $8.75   

[4622 rows x 6 columns]

### 3. What are the most popular 3 items?


In [147]:
pd.crosstab(orders.item_name, orders.quantity.sum())

col_0                                  4972
item_name                                  
6 Pack Soft Drink                        54
Barbacoa Bowl                            66
Barbacoa Burrito                         91
Barbacoa Crispy Tacos                    11
Barbacoa Salad Bowl                      10
Barbacoa Soft Tacos                      25
Bottled Water                           162
Bowl                                      2
Burrito                                   6
Canned Soda                             104
Canned Soft Drink                       301
Carnitas Bowl                            68
Carnitas Burrito                         59
Carnitas Crispy Tacos                     7
Carnitas Salad                            1
Carnitas Salad Bowl                       6
Carnitas Soft Tacos                      40
Chicken Bowl                            726
Chicken Burrito                         553
Chicken Crispy Tacos                     47
Chicken Salad                             9
Chicken Salad Bowl                      110
Chicken Soft Tacos                      115
Chips                                   211
Chips and Fresh Tomato Salsa            110
Chips and Guacamole                     479
Chips and Mild Fresh Tomato Salsa         1
Chips and Roasted Chili Corn Salsa       22
Chips and Roasted Chili-Corn Salsa       18
Chips and Tomatillo Green Chili Salsa    43
Chips and Tomatillo Red Chili Salsa      48
Chips and Tomatillo-Green Chili Salsa    31
Chips and Tomatillo-Red Chili Salsa      20
Crispy Tacos                              2
Izze                                     20
Nantucket Nectar                         27
Salad                                     2
Side of Chips                           101
Steak Bowl                              211
Steak Burrito                           368
Steak Crispy Tacos                       35
Steak Salad                               4
Steak Salad Bowl                         29
Steak Soft Tacos                         55
Veggie Bowl                              85
Veggie Burrito                           95
Veggie Crispy Tacos                       1
Veggie Salad                              6
Veggie Salad Bowl                        18
Veggie Soft Tacos                         7

In [148]:
num_item_ordered = orders.groupby('item_name').quantity.agg(['sum'])
num_item_ordered.nlargest(3, 'sum')

sum
item_name               
Chicken Bowl         761
Chicken Burrito      591
Chips and Guacamole  506

### 4. Which item has produced the most revenue?



In [154]:
num_item_ordered.rename(columns = {'item_name':'item_name','sum':'total'}, inplace = True)
num_item_ordered

total
item_name                                   
6 Pack Soft Drink                         55
Barbacoa Bowl                             66
Barbacoa Burrito                          91
Barbacoa Crispy Tacos                     12
Barbacoa Salad Bowl                       10
Barbacoa Soft Tacos                       25
Bottled Water                            211
Bowl                                       4
Burrito                                    6
Canned Soda                              126
Canned Soft Drink                        351
Carnitas Bowl                             71
Carnitas Burrito                          60
Carnitas Crispy Tacos                      8
Carnitas Salad                             1
Carnitas Salad Bowl                        6
Carnitas Soft Tacos                       40
Chicken Bowl                             761
Chicken Burrito                          591
Chicken Crispy Tacos                      50
Chicken Salad                              9
Chicken Salad Bowl                       123
Chicken Soft Tacos                       120
Chips                                    230
Chips and Fresh Tomato Salsa             130
Chips and Guacamole                      506
Chips and Mild Fresh Tomato Salsa          1
Chips and Roasted Chili Corn Salsa        23
Chips and Roasted Chili-Corn Salsa        18
Chips and Tomatillo Green Chili Salsa     45
Chips and Tomatillo Red Chili Salsa       50
Chips and Tomatillo-Green Chili Salsa     33
Chips and Tomatillo-Red Chili Salsa       25
Crispy Tacos                               2
Izze                                      20
Nantucket Nectar                          29
Salad                                      2
Side of Chips                            110
Steak Bowl                               221
Steak Burrito                            386
Steak Crispy Tacos                        36
Steak Salad                                4
Steak Salad Bowl                          31
Steak Soft Tacos                          56
Veggie Bowl                               87
Veggie Burrito                            97
Veggie Crispy Tacos                        1
Veggie Salad                               6
Veggie Salad Bowl                         18
Veggie Soft Tacos                          8

In [165]:
item_revenue = num_item_ordered.total * orders.item_price
type(item_revenue)

pandas.core.series.Series

In [173]:
num_item_ordered = orders.groupby('item_name').quantity.agg(['sum'])
num_item_ordered.nlargest(1, 'sum')

sum
item_name        
Chicken Bowl  761

### 5. Join the employees and titles DataFrames together.


In [187]:
employee_titles = pd.merge(employees, titles, left_on='emp_no', right_on='emp_no', how='inner')
employee_titles

emp_no  birth_date first_name last_name gender   hire_date  \
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26   
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21   
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28   
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
4        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
...        ...         ...        ...       ...    ...         ...   
443303  499997  1961-08-03    Berhard    Lenart      M  1986-04-21   
443304  499997  1961-08-03    Berhard    Lenart      M  1986-04-21   
443305  499998  1956-09-05   Patricia   Breugel      M  1993-10-13   
443306  499998  1956-09-05   Patricia   Breugel      M  1993-10-13   
443307  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30   

                  title   from_date     to_date  
0       Senior Engineer  1986-06-26  9999-01-01  
1                 Staff  1996-08-03  9999-01-01  
2       Senior Engineer  1995-12-03  9999-01-01  
3              Engineer  1986-12-01  1995-12-01  
4       Senior Engineer  1995-12-01  9999-01-01  
...                 ...         ...         ...  
443303         Engineer  1987-08-30  1992-08-29  
443304  Senior Engineer  1992-08-29  9999-01-01  
443305     Senior Staff  1998-12-27  9999-01-01  
443306            Staff  1993-12-27  1998-12-27  
443307         Engineer  1997-11-30  9999-01-01  

[443308 rows x 9 columns]

### 6. For each title, find the hire date of the employee that was hired most recently with that title.


In [188]:
employee_titles.groupby(by = 'title').hire_date.max()

title
Assistant Engineer    1999-12-24
Engineer              2000-01-28
Manager               1992-02-05
Senior Engineer       2000-01-01
Senior Staff          2000-01-13
Staff                 2000-01-12
Technique Leader      1999-12-31
Name: hire_date, dtype: object

### 7. Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)

In [191]:
df = pd.read_sql('SELECT * FROM employees JOIN titles USING(emp_no) JOIN dept_emp USING(emp_no) JOIN departments USING(dept_no)', get_db_url('employees'))
df

dept_no  emp_no  birth_date first_name last_name gender   hire_date  \
0         d009   10011  1953-11-07       Mary     Sluis      F  1990-01-22   
1         d009   10038  1960-07-20       Huan     Lortz      M  1989-09-20   
2         d009   10038  1960-07-20       Huan     Lortz      M  1989-09-20   
3         d009   10049  1961-04-24      Basil    Tramer      F  1992-05-04   
4         d009   10049  1961-04-24      Basil    Tramer      F  1992-05-04   
...        ...     ...         ...        ...       ...    ...         ...   
489898    d007  499986  1952-07-22     Nathan     Ranta      F  1985-08-11   
489899    d007  499986  1952-07-22     Nathan     Ranta      F  1985-08-11   
489900    d007  499987  1961-09-05      Rimli    Dusink      F  1998-09-20   
489901    d007  499988  1962-09-28   Bangqing   Kleiser      F  1986-06-06   
489902    d007  499988  1962-09-28   Bangqing   Kleiser      F  1986-06-06   

               title   from_date     to_date   from_date     to_date  \
0              Staff  1990-01-22  1996-11-09  1990-01-22  1996-11-09   
1       Senior Staff  1996-09-20  9999-01-01  1989-09-20  9999-01-01   
2              Staff  1989-09-20  1996-09-20  1989-09-20  9999-01-01   
3       Senior Staff  2000-05-04  9999-01-01  1992-05-04  9999-01-01   
4              Staff  1992-05-04  2000-05-04  1992-05-04  9999-01-01   
...              ...         ...         ...         ...         ...   
489898  Senior Staff  1992-08-11  9999-01-01  1985-08-11  9999-01-01   
489899         Staff  1985-08-11  1992-08-11  1985-08-11  9999-01-01   
489900         Staff  1999-12-21  9999-01-01  1999-12-21  9999-01-01   
489901  Senior Staff  1997-07-25  2001-10-09  1988-07-25  2001-10-09   
489902         Staff  1988-07-25  1997-07-25  1988-07-25  2001-10-09   

               dept_name  
0       Customer Service  
1       Customer Service  
2       Customer Service  
3       Customer Service  
4       Customer Service  
...                  ...  
489898             Sales  
489899             Sales  
489900             Sales  
489901             Sales  
489902             Sales  

[489903 rows x 13 columns]